File: internal/models/application/application.go
// Package application 定义应用管理相关的数据模型
// 本文件包含应用管理相关的模型定义
package application

import (
	"time"
)

// ==================== 应用管理模型 ====================

// Organization 组织
type Organization struct {
	ID          uint      `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
	Name        string    `gorm:"size:100;not null;uniqueIndex" json:"name"`
	DisplayName string    `gorm:"size:200" json:"display_name"`
	Description string    `gorm:"type:text" json:"description"`
	Owner       string    `gorm:"size:100" json:"owner"`
	Status      string    `gorm:"size:20;default:'active'" json:"status"`
}

func (Organization) TableName() string { return "organizations" }

// Project 项目
type Project struct {
	ID             uint      `gorm:"primarykey" json:"id"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
	OrganizationID uint      `gorm:"not null;index" json:"organization_id"`
	Name           string    `gorm:"size:100;not null" json:"name"`
	DisplayName    string    `gorm:"size:200" json:"display_name"`
	Description    string    `gorm:"type:text" json:"description"`
	Owner          string    `gorm:"size:100" json:"owner"`
	Status         string    `gorm:"size:20;default:'active'" json:"status"`
	OrgName        string    `gorm:"-" json:"org_name,omitempty"`
}

func (Project) TableName() string { return "projects" }

// EnvDefinition 自定义环境定义
type EnvDefinition struct {
	ID          uint      `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time `json:"created_at"`
	Name        string    `gorm:"size:50;not null;uniqueIndex" json:"name"`
	DisplayName string    `gorm:"size:100" json:"display_name"`
	SortOrder   int       `gorm:"default:0" json:"sort_order"`
	Color       string    `gorm:"size:20;default:'blue'" json:"color"`
}

func (EnvDefinition) TableName() string { return "env_definitions" }

// Application 应用模型
// 存储应用的基本信息和关联配置
type Application struct {
	ID                uint      `gorm:"primarykey" json:"id"`
	CreatedAt         time.Time `json:"created_at"`
	UpdatedAt         time.Time `json:"updated_at"`
	Name              string    `gorm:"size:100;not null;uniqueIndex" json:"name"` // 应用名称，唯一
	DisplayName       string    `gorm:"size:200" json:"display_name"`              // 显示名称
	Description       string    `gorm:"type:text" json:"description"`              // 描述
	OrganizationID    *uint     `gorm:"index" json:"organization_id"`              // 所属组织
	ProjectID         *uint     `gorm:"index" json:"project_id"`                   // 所属项目
	GitRepo           string    `gorm:"size:500" json:"git_repo"`                  // Git 仓库地址
	Language          string    `gorm:"size:50" json:"language"`                   // 开发语言
	Framework         string    `gorm:"size:50" json:"framework"`                  // 框架
	Team              string    `gorm:"size:100;index" json:"team"`                // 所属团队
	Owner             string    `gorm:"size:100" json:"owner"`                     // 负责人
	OrgName           string    `gorm:"-" json:"org_name,omitempty"`
	ProjectName       string    `gorm:"-" json:"project_name,omitempty"`
	Status            string    `gorm:"size:20;default:'active'" json:"status"` // 状态
	NotifyPlatform    string    `gorm:"size:50" json:"notify_platform"`         // 通知平台
	NotifyAppID       *uint     `gorm:"index" json:"notify_app_id"`             // 通知应用ID
	NotifyReceiveID   string    `gorm:"size:200" json:"notify_receive_id"`      // 通知接收者ID
	NotifyReceiveType string    `gorm:"size:50" json:"notify_receive_type"`     // 接收者类型
	CreatedBy         *uint     `gorm:"index" json:"created_by"`
}

// TableName 指定表名
func (Application) TableName() string {
	return "applications"
}

// ApplicationRepoBinding 应用与标准 Git 仓库的绑定关系。
// 一个应用可绑定多个仓库，但同一应用只允许一个默认主仓库。
type ApplicationRepoBinding struct {
	ID            uint      `gorm:"primarykey" json:"id"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
	ApplicationID uint      `gorm:"not null;index:idx_app_repo_binding_app;uniqueIndex:uk_app_repo_binding" json:"application_id"`
	GitRepoID     uint      `gorm:"not null;index:idx_app_repo_binding_repo;uniqueIndex:uk_app_repo_binding" json:"git_repo_id"`
	Role          string    `gorm:"size:30;not null;default:'primary'" json:"role"`
	IsDefault     bool      `gorm:"not null;default:false;index:idx_app_repo_binding_default" json:"is_default"`
	CreatedBy     *uint     `gorm:"index" json:"created_by"`

	RepoName      string `gorm:"-" json:"repo_name,omitempty"`
	RepoURL       string `gorm:"-" json:"repo_url,omitempty"`
	RepoProvider  string `gorm:"-" json:"repo_provider,omitempty"`
	DefaultBranch string `gorm:"-" json:"default_branch,omitempty"`
}

func (ApplicationRepoBinding) TableName() string { return "application_repo_bindings" }

// ApplicationEnv 应用环境配置
// 存储应用在不同环境下的配置
type ApplicationEnv struct {
	ID                  uint      `gorm:"primarykey" json:"id"`
	CreatedAt           time.Time `json:"created_at"`
	UpdatedAt           time.Time `json:"updated_at"`
	ApplicationID       uint      `gorm:"column:app_id;not null;index" json:"application_id"`              // 应用ID
	EnvName             string    `gorm:"size:50;not null" json:"env_name"`                                // 环境名称
	Branch              string    `gorm:"size:100" json:"branch"`                                          // 业务代码分支
	GitOpsRepoID        *uint     `gorm:"column:gitops_repo_id;index" json:"gitops_repo_id"`               // GitOps 部署仓库ID
	ArgoCDApplicationID *uint     `gorm:"column:argocd_application_id;index" json:"argocd_application_id"` // ArgoCD 应用ID
	GitOpsBranch        string    `gorm:"column:gitops_branch;size:200" json:"gitops_branch"`              // GitOps 目标分支
	GitOpsPath          string    `gorm:"column:gitops_path;size:500" json:"gitops_path"`                  // GitOps 部署目录
	HelmChartPath       string    `gorm:"size:500" json:"helm_chart_path"`                                 // Helm Chart 路径
	HelmValuesPath      string    `gorm:"size:500" json:"helm_values_path"`                                // Helm values 文件路径
	HelmReleaseName     string    `gorm:"size:200" json:"helm_release_name"`                               // Helm Release 名称
	K8sClusterID        *uint     `gorm:"index" json:"k8s_cluster_id"`                                     // K8s 集群ID
	K8sNamespace        string    `gorm:"size:100" json:"k8s_namespace"`                                   // K8s 命名空间
	K8sDeployment       string    `gorm:"size:200" json:"k8s_deployment"`                                  // K8s Deployment
	Replicas            int       `gorm:"default:1" json:"replicas"`                                       // 副本数
	CPURequest          string    `gorm:"size:50" json:"cpu_request"`                                      // CPU request
	CPULimit            string    `gorm:"size:50" json:"cpu_limit"`                                        // CPU limit
	MemoryRequest       string    `gorm:"size:50" json:"memory_request"`                                   // Memory request
	MemoryLimit         string    `gorm:"size:50" json:"memory_limit"`                                     // Memory limit
	Config              string    `gorm:"type:text" json:"config"`                                         // 环境特定配置
}

// TableName 指定表名
func (ApplicationEnv) TableName() string {
	return "application_envs"
}

// ApplicationReadinessCheck records the latest delivery readiness snapshot for an application.
type ApplicationReadinessCheck struct {
	ID            uint      `gorm:"primarykey" json:"id"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
	ApplicationID uint      `gorm:"not null;uniqueIndex:idx_app_readiness_app_key;index" json:"application_id"`
	CheckKey      string    `gorm:"size:80;not null;uniqueIndex:idx_app_readiness_app_key" json:"check_key"`
	Title         string    `gorm:"size:120;not null" json:"title"`
	Description   string    `gorm:"size:500" json:"description"`
	Status        string    `gorm:"size:20;not null;index" json:"status"`
	Severity      string    `gorm:"size:20;not null;default:'info';index" json:"severity"`
	Path          string    `gorm:"size:300" json:"path"`
	Score         int       `gorm:"default:0" json:"score"`
	Level         string    `gorm:"size:30;not null;default:'not_ready'" json:"level"`
	CheckedAt     time.Time `gorm:"not null;index" json:"checked_at"`
}

func (ApplicationReadinessCheck) TableName() string {
	return "application_readiness_checks"
}

File: web/src/services/application.ts
import request from './api'
import type { ApiResponse } from '../types'

export interface Application {
  id?: number
  created_at?: string
  updated_at?: string
  name: string
  display_name?: string
  description?: string
  organization_id?: number
  project_id?: number
  git_repo?: string
  language?: string
  framework?: string
  team?: string
  owner?: string
  org_name?: string
  project_name?: string
  status: string
  notify_platform?: string
  notify_app_id?: number
  notify_receive_id?: string
  notify_receive_type?: string
}

export interface ApplicationEnv {
  id?: number
  application_id: number
  env_name: string
  branch?: string
  gitops_repo_id?: number
  argocd_application_id?: number
  gitops_branch?: string
  gitops_path?: string
  helm_chart_path?: string
  helm_values_path?: string
  helm_release_name?: string
  k8s_cluster_id?: number
  k8s_namespace?: string
  k8s_deployment?: string
  replicas?: number
  cpu_request?: string
  cpu_limit?: string
  memory_request?: string
  memory_limit?: string
  config?: string
}

export interface DeliveryRecord {
  id: number
  created_at: string
  application_id: number
  app_name: string
  env_name: string
  version: string
  branch: string
  commit_id: string
  image_tag?: string
  deploy_type: string
  deploy_method?: string
  status: string
  duration: number
  error_msg?: string
  need_approval?: boolean
  approval_chain_id?: number
  operator: string
  started_at?: string
  finished_at?: string
}

export interface ApplicationRepoBinding {
  id: number
  application_id: number
  git_repo_id: number
  role: string
  is_default: boolean
  created_at?: string
  updated_at?: string
  repo_name?: string
  repo_url?: string
  repo_provider?: string
  default_branch?: string
}

export interface AppStats {
  app_count: number
  team_stats: { name: string; count: number }[]
  lang_stats: { name: string; count: number }[]
  today_deliveries: number
  week_deliveries: number
  success_rate: number
}

export interface ApplicationReadinessCheck {
  key: string
  title: string
  description: string
  status: 'pass' | 'missing' | string
  severity: 'info' | 'low' | 'medium' | 'high' | string
  path?: string
}

export interface ApplicationReadinessAction {
  key: string
  title: string
  path: string
  weight: number
}

export interface ApplicationReadiness {
  application_id: number
  application_name: string
  score: number
  level: string
  completed: number
  total: number
  checks: ApplicationReadinessCheck[]
  next_actions: ApplicationReadinessAction[]
  generated_at: string
}

export interface ApplicationOnboardingRequest {
  application_id?: number
  app: Partial<Application>
  repo?: {
    git_repo_id?: number
    name?: string
    url?: string
    provider?: string
    default_branch?: string
    role?: string
    is_default?: boolean
  }
  env?: Partial<ApplicationEnv>
  pipeline?: {
    pipeline_id?: number
    create?: boolean
    name?: string
    description?: string
    env?: string
    source_template_id?: number
    git_repo_id?: number
    git_branch?: string
  }
}

export interface ApplicationOnboardingResponse {
  application_id: number
  application_name: string
  created: boolean
  repo_binding_id?: number
  git_repo_id?: number
  env_id?: number
  pipeline_id?: number
  updated_sections: string[]
  readiness?: ApplicationReadiness
}

export interface DeployWindowStatus {
  in_window: boolean
  message: string
  next_window?: string
}

export interface DeployLockStatus {
  locked: boolean
  locked_by?: string
  locked_at?: string
}

export interface ApprovalRequiredStatus {
  required: boolean
  approvers?: string[]
}

export const getDeployWindowStatus = (appId: number, envName: string): Promise<ApiResponse<DeployWindowStatus>> => {
  return request.get(`/approval/windows/check`, {
    params: { app_id: appId, env: envName },
    skipErrorToast: true,
  })
}

export const getDeployLockStatus = (appId: number, envName: string): Promise<ApiResponse<DeployLockStatus>> => {
  return request.get(`/deploy/locks/check`, {
    params: { app_id: appId, env: envName },
    skipErrorToast: true,
  })
}

export const checkApprovalRequired = (appId: number, envName: string): Promise<ApiResponse<ApprovalRequiredStatus>> => {
  return request.get(`/approval/check`, {
    params: { app_id: appId, env: envName },
    skipErrorToast: true,
  })
}

export const applicationApi = {
  // 应用管理
  list: (params?: { page?: number; page_size?: number; name?: string; team?: string; status?: string; language?: string; organization_id?: number; project_id?: number }): Promise<ApiResponse<{ list: Application[]; total: number }>> => {
    return request.get('/app', { params })
  },

  get: (id: number): Promise<ApiResponse<{ app: Application; envs: ApplicationEnv[]; repo_bindings?: ApplicationRepoBinding[]; default_repo_binding?: ApplicationRepoBinding }>> => {
    return request.get(`/app/${id}`)
  },

  create: (data: Partial<Application>): Promise<ApiResponse<Application>> => {
    return request.post('/app', data)
  },

  update: (id: number, data: Partial<Application>): Promise<ApiResponse<Application>> => {
    return request.put(`/app/${id}`, data)
  },

  delete: (id: number): Promise<ApiResponse> => {
    return request.delete(`/app/${id}`)
  },

  saveOnboarding: (data: ApplicationOnboardingRequest): Promise<ApiResponse<ApplicationOnboardingResponse>> => {
    return request.post('/app/onboarding', data)
  },

  // 环境管理
  listEnvs: (appId: number): Promise<ApiResponse<ApplicationEnv[]>> => {
    return request.get(`/app/${appId}/envs`)
  },

  getReadiness: (appId: number): Promise<ApiResponse<ApplicationReadiness>> => {
    return request.get(`/app/${appId}/readiness`, { skipErrorToast: true })
  },

  refreshReadiness: (appId: number): Promise<ApiResponse<ApplicationReadiness>> => {
    return request.post(`/app/${appId}/readiness/run`, undefined, { skipErrorToast: true })
  },

  listRepoBindings: (appId: number): Promise<ApiResponse<ApplicationRepoBinding[]>> => {
    return request.get(`/app/${appId}/repo-bindings`)
  },

  bindRepo: (appId: number, data: { git_repo_id: number; role?: string; is_default?: boolean }): Promise<ApiResponse<ApplicationRepoBinding>> => {
    return request.post(`/app/${appId}/repo-bindings`, data)
  },

  setDefaultRepoBinding: (appId: number, bindingId: number): Promise<ApiResponse> => {
    return request.put(`/app/${appId}/repo-bindings/${bindingId}/default`)
  },

  deleteRepoBinding: (appId: number, bindingId: number): Promise<ApiResponse> => {
    return request.delete(`/app/${appId}/repo-bindings/${bindingId}`)
  },

  createEnv: (appId: number, data: Partial<ApplicationEnv>): Promise<ApiResponse<ApplicationEnv>> => {
    return request.post(`/app/${appId}/envs`, data)
  },

  updateEnv: (appId: number, envId: number, data: Partial<ApplicationEnv>): Promise<ApiResponse<ApplicationEnv>> => {
    return request.put(`/app/${appId}/envs/${envId}`, data)
  },

  deleteEnv: (appId: number, envId: number): Promise<ApiResponse> => {
    return request.delete(`/app/${appId}/envs/${envId}`)
  },

  // 交付记录
  listDeliveryRecords: (appId: number, params?: { page?: number; page_size?: number; env_name?: string; status?: string }): Promise<ApiResponse<{ list: DeliveryRecord[]; total: number }>> => {
    return request.get(`/app/${appId}/delivery-records`, { params })
  },

  listAllDeliveryRecords: (params?: { page?: number; page_size?: number; app_name?: string; env?: string; status?: string }): Promise<ApiResponse<{ list: DeliveryRecord[]; total: number }>> => {
    return request.get('/app/delivery-records', { params })
  },

  // 统计
  getStats: (): Promise<ApiResponse<AppStats>> => {
    return request.get('/app/stats')
  },

  getTeams: (): Promise<ApiResponse<string[]>> => {
    return request.get('/app/teams')
  },

  getDeployWindowStatus,
  getDeployLockStatus,
  checkApprovalRequired,
}
